main.go

```diff
@@ -132,15 +133,64 @@ type internalTracingConfig struct {
 	samplingFraction float64
 }
 
+type tenant struct {
+	Name string `json:"name"`
+	ID   string `json:"id"`
+	OIDC *struct {
+		ClientID      string `json:"clientID"`
+		ClientSecret  string `json:"clientSecret"`
+		GroupClaim    string `json:"groupClaim"`
+		IssuerRawCA   []byte `json:"issuerCA"`
+		IssuerCAPath  string `json:"issuerCAPath"`
+		issuerCA      *x509.Certificate
+		IssuerURL     string `json:"issuerURL"`
+		RedirectURL   string `json:"redirectURL"`
+		UsernameClaim string `json:"usernameClaim"`
+	} `json:"oidc"`
+	MTLS *struct {
+		RawCA  []byte `json:"ca"`
+		CAPath string `json:"caPath"`
+		cas    []*x509.Certificate
+	} `json:"mTLS"`
+	OPA *struct {
+		Query      string   `json:"query"`
+		Paths      []string `json:"paths"`
+		URL        string   `json:"url"`
+		authorizer rbac.Authorizer
+	} `json:"opa"`
+	RateLimits []*struct {
+		Endpoint string   `json:"endpoint"`
+		Limit    int      `json:"limit"`
+		Window   duration `json:"window"`
+	} `json:"rateLimits"`
+}
+
+type tenantsConfig struct {
+	Tenants         []*tenant `json:"tenants"`
+	rateLimitClient *ratelimit.Client
+	authorizer      rbac.Authorizer
+	ins             signalhttp.HandlerInstrumenter
+	reg             *prometheus.Registry
+	logger          log.Logger
+}
+
 //nolint:funlen,gocyclo,gocognit
 func main() {
 	cfg, err := parseFlags()
 	if err != nil {
 		stdlog.Fatalf("parse flag: %v", err)
 	}
 
-	logger := logger.NewLogger(cfg.logLevel, cfg.logFormat, cfg.debug.name)
-	defer level.Info(logger).Log("msg", "exiting")
+	// Load all the command line configuration into tenantsConfig struct.
+	tCfg := tenantsConfig{}
+	loadTenantConfigs(&cfg, &tCfg)
```

Comment:
This is a strange pattern reminding on old C days. Prefer to return the full-backed tenants config, e.g.:
```go
tCfg := loadTenantConfig(&cfg)
```